### PR TITLE
Add weighting option to CopyNet

### PR DESCRIPTION
**Pull request by epwalsh**
This is just a minor, general improvement to CopyNet. It adds the option to weight the loss contributions from individual instances when calculating the batch loss.

### CODE REVIEW

File: allennlp_models/generation/dataset_readers/copynet_seq2seq.py [hunk 1]
```diff
@@ -194,6 +202,9 @@ def text_to_instance(
 
         fields_dict["metadata"] = MetadataField(meta_fields)
 
+        if weight is not None:
+            fields_dict["weight"] = ArrayField(np.array([float(weight)], dtype=np.single))
```
Comment (dirkgr) [inline allennlp_models/generation/dataset_readers/copynet_seq2seq.py:206]: Just make it a `TensorField`, and use Torch methods to create the tensors. `ArrayField` is the old way.

File: allennlp_models/generation/models/copynet_seq2seq.py [hunk 2]
```diff
@@ -527,7 +536,13 @@ def _forward_loss(
...
         log_likelihood = (log_likelihoods * target_mask).sum(dim=-1)
         # The loss is the negative log-likelihood, averaged over the batch.
-        loss = -log_likelihood.sum() / batch_size
+        if weight is None:
+            loss = -log_likelihood.sum() / batch_size
+        else:
+            # shape: (batch_size,)
+            if len(weight.shape) > 1:
+                weight = weight.squeeze()
+            loss = -(weight * log_likelihood).sum() / batch_size
```
Comment (dirkgr) [inline allennlp_models/generation/models/copynet_seq2seq.py:545]: Shouldn't this be `/ weight.sum()`?

**Review by dirkgr — CHANGES_REQUESTED** [review]

File: allennlp_models/generation/models/copynet_seq2seq.py (hunk 2 above)
Comment (epwalsh) [reply allennlp_models/generation/models/copynet_seq2seq.py:545]: I've gone back and forth on this.

If you use `weight.sum()` to normalize, then the weighting is only relative to each batch, which is probably *not* what you want.

For example, let's say we normalize by `weight.sum()` and your weights range from 0.5 - 1.0. If you have a batch that contains only instances with weights of 0.5, then this will give you the same result as if they all had weights of 1.0.

Comment (dirkgr) [reply allennlp_models/generation/models/copynet_seq2seq.py:545]: Hmm, fair enough. I would expect that setting all the weights to 1000 should be the same as setting them all to 0.001. But to get that behavior, and also the behavior you want, we would have to sum up all the weights in the dataset before processing a single batch, and then scale each batch accordingly. That's not practical. So let's leave it like this then.

**Review by dirkgr — APPROVED** [review]
